### PR TITLE
fix bug in the way weighted objective is computed for time series mask

### PR DESCRIPTION
**Pull request by udibr**
this looks like a critical bug for anyone working with time series. So I'm surprised it didn't surfaced before. Maybe I'm missing something?

This PR does not fix an additional bug in this function which happens if using time series weights and mask together.

Also the line
        if weights is not None:
which appears just *after* weights is used looks redundant (at best)

### CODE REVIEW

**Comment by fchollet** [comment]
The mask always has the same shape as the output of the loss function, by design. How is there a bug?

**Comment by udibr** [comment]
When computing the loss you want to take only the time steps for which mask is ,
this is why you did `score_array *= mask` you then want to average their loss but you cant just use `mean` over the entire sample because you don't want to average in the 0 loss from steps in which mask was 0. The original code therefore divided by the mask of mean before doing a regular mean to cancel this effect. But this is wrong because each sample can have a different number of 1s in its row in the mask.

You therefore need to divide each sample/row by a different factor depending on how much 1s it has in the mask. The mask shape does not change because I'm contracting and then expanding back.

There remain a bug in what I did if an entire sample/row has all its mask bit set to 0.
In this case the loss of the sample should be removed from the mean over the entire batch.

Also the fact that not all bits are turned on in the mask and that it is different for different samples in the batch should affect how weights are being used. I didnt touch that.

Also there is an if statement that clearly has no effect because it checks if weights is not None just *after* using weights.

As some famous person said, it is better to re-write this entire method than to slowly refactor it

**Comment by udibr** [comment]
Ok I did misunderstood... sorry.
But the code had a bug:
It does not make sense to give a different weight to an entire batch according mean-mask
instead what you should do is first find out if any of the samples in the batch was unmasked.
In the case of time series this will happen in the rare case in which all the steps of a single sample had mask set to 0.

you should then compute the mean of the entire batch but only on the samples which were masked. In most cases this will always be all samples so this bug had very little effect on results in the past.

**Comment by fchollet** [comment]
Why did you close it? Was there a actually problem, or not?

**Comment by udibr** [comment]
Keras is OK because it computes the average loss per word.
When using a mask the average loss per word is different from the average loss per sequence (in which you sum the loss of all the words in the sequence) and this made me think there was a bug.
